### ISSUE TITLE
Publish Hazelcast testing classes to Maven central

### ISSUE THREAD

**Issue by markathomas** [hazelcast/hazelcast#5017]
I'd really like to see Hazelcast publish it's unit test framework to Maven central.  For those of us writing unit tests for our own Hazelcast code the examples and boilerplate code already created by the Hazelcast team would be very valuable.

**Comment by jerrinot** [comment]
Hi Mark,

do you mean our test support classes? I believe it's already available in Maven central. You just need to add a classifier "tests". 
```xml
        <dependency>
            <groupId>com.hazelcast</groupId>
            <artifactId>hazelcast</artifactId>
            <version>3.4.2</version>
            <classifier>tests</classifier>
        </dependency>
```
Please bare in mind API of the test support is not officially supported and it can change without any notice.

**Comment by markathomas** [comment]
Ah yes, thanks for pointing that out.  Looks like the test sources are
still missing though?

Regards,

Mark Thomas
spatialguru.net@gmail.com
205.529.9013

"Commit to the Lord whatever you do,
    and your plans will succeed." - Proverbs 16:3

**Comment by jerrinot** [comment]
you are right! let's see what we can do about it. 
thanks for reporting it.

**Comment by jerrinot** [comment]
@mesutcelik: Is it possible to enhance our release script to include test sources as well?

**Comment by markathomas** [comment]
great.  thanks a bunch!

Regards,

Mark Thomas
spatialguru.net@gmail.com
205.529.9013

"Commit to the Lord whatever you do,
    and your plans will succeed." - Proverbs 16:3

**Comment by mesutcelik** [comment]
It is doable by updating ```maven-source-plugin``` inside hazelcast module by including test sources.

```
            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-source-plugin</artifactId>
                <version>${maven.source.plugin.version}</version>
                <executions>
                    <execution>
                        <id>attach-sources</id>
                        <phase>verify</phase>
                        <goals>
                            <goal>jar-no-fork</goal>
                            <goal>test-jar-no-fork</goal>
                        </goals>
                    </execution>
                </executions>
            </plugin>
```

and then maven deploy during release will send sources to maven central.

**Comment by pveentjer** [comment]
I think we should prevent publishing the testing framework. Because at that point in time it becomes a public API and then we need to follow certain practices like deprecation etc.

**Comment by gAmUssA** [comment]
@pveentjer Personally, I think it's great idea to publish test framework. it's huge help for TechOps team so we can be more productive working with clients' problems - we can talk on the same language with our dev using test Hazelcast test framework.

**Comment by markathomas** [comment]
I agree with Viktor. Publishing the test classes makes it easily available
to all and provides  a ready means by which to reproduce bugs as test cases
based on code the Hazelcast team is familiar with.  Additionally, it shows
common patterns for successful testing which will increase developer
confidence and thus wider adoption of Hazelcast.

**Comment by pveentjer** [comment]
I understand the need for it. But I'm very worried about the backwards compatibility issues. So I want to expose the least amount as possible. And you can get the test library; only not from a public repo. That is it.

**Comment by dbrimley** [comment]
I'm in favour of publishing the source code to the tests.
